### DISCUSSION TITLE
[@types/react] forwardRef loses type parameters (higher order function type inference)

### DISCUSSION THREAD

**Discussion by AndyOGo** [DefinitelyTyped/DefinitelyTyped#69029] [Issues with a @types package]
`forwardRef` doesn't support [higher order function type inference](https://github.com/microsoft/TypeScript/pull/30215) which breaks generics.

Details about the problem can be found at that SO entry.
> We only make **higher order function type inferences** when the source and target types are both pure function types, i.e. types with a single call signature and no other members [[1]](https://github.com/microsoft/TypeScript/issues/30650#issuecomment-478258803), [[2]](https://github.com/microsoft/TypeScript/issues/30650#issuecomment-486680485).
https://stackoverflow.com/questions/58469229/react-with-typescript-generics-while-using-react-forwardref/58473012

---

My local workaround by augmenting types with a declaration file.
- [`displayName`](https://legacy.reactjs.org/docs/forwarding-refs.html#displaying-a-custom-name-in-devtools) works,
- but `propTypes` and `defaultProps` aren't.

```ts
import React from "react"

// Redecalare forwardRef
declare module "react" {
  function ComponentToRender<T, P>(props: P, ref: React.ForwardedRef<T>): React.ReactElement | null;
  ComponentToRender.displayName as string;

  function ForwardedComponent<T, P>(props: P & React.RefAttributes<T>): React.ReactElement | null;
  ForwardedComponent.displayName as string;

  function forwardRef<T, P = {}>(
    render: Render<T, P>
  ): ForwardedComponent<T, P>;
}
```

[a comment by an automated account is not shown]

**Comment by AndyOGo** [comment]
Probably related #52216, #42691, #30451

**Comment by eps1lon** [comment]
The original type definition preceeds the linked TypeScript feature. Feel free to investigate how we can make this work and propose changes via PR.

**Reply by AndyOGo** [reply]
I updated my  description.
I could only get [`displayName`](https://legacy.reactjs.org/docs/forwarding-refs.html#displaying-a-custom-name-in-devtools) working.
But `propsTypes` and `defaultProps` aren't.

Anyway `defaultProps` seem to have other issues
https://stackoverflow.com/questions/74154313/react-typescript-forwardref-with-defaultprops

**Reply by eps1lon** [reply]
`defaultProps` do not work on the render functions passed to `forwardRef`. Same for `propTypes`. The value you pass to `forwardRef` is not a component but a special "render function".

**Reply by AndyOGo** [reply]
I'm glad you say this.

Because the types of `ForwardRefRenderFunction` state otherwise:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f6b37733d6a66dc9794fffebf9a4757008b5faa1/types/react/index.d.ts#L1204-L1231

**Reply by eps1lon** [reply]
> Because the types of ForwardRefRenderFunction state otherwise:

Why do you think the types contradict what I said? `propTypes` and `defaultProps` are not supported hence the `never` to highlight accidentally setting those.

**Reply by AndyOGo** [reply]
Sorry, my bad.
I just noticed they are typed as `never`.

**Reply by AndyOGo** [reply]
The returned component is typed as `ForwardRefExoticComponent` and has these props though.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f6b37733d6a66dc9794fffebf9a4757008b5faa1/types/react/index.d.ts#L1521-L1524

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f6b37733d6a66dc9794fffebf9a4757008b5faa1/types/react/index.d.ts#L1553-L1555

---

FYI: I double checked the source and can't quickly find them being returned.

https://github.com/facebook/react/blob/main/packages/react/src/ReactForwardRef.js

**Reply by eps1lon** [reply]
The returned type accepts these, yes. But you linked the input type i.e. `ForwardRefRenderFunction` which doesn't accept propTypes and defaultProps

**Reply by AndyOGo** [reply]
Exactly.
In `react`'s source I can't find them being returned.
https://github.com/facebook/react/blob/b09e102ff1e2aaaf5eb6585b04609ac7ff54a5c8/packages/react/src/ReactForwardRef.js#L76

**Reply by eps1lon** [reply]
There wouldn't be anything to return. But you can add them later and React would recognize them, hence the types for them: https://codesandbox.io/p/sandbox/quizzical-sea-jhykwh

**Reply by AndyOGo** [reply]
True, I didn't think about that.
Thx for pointing it out.

Unfortunately I'm afraid supporting those types makes higher order function type inference impossible then.

**Reply by eps1lon** [reply]
What if we dont leverage a special interface but return a function that's intersected with an object with propTypes+defaultProps?

**Reply by AndyOGo** [reply]
That's worth trying.
I wonder if TS still bails-out of higher order function type inference.

**Reply by mattpocock** [reply]
TS bails out when it encounters other keys on the function.

In other words, `displayName` and `defaultProps` are preventing it from working.

https://www.totaltypescript.com/forwardref-with-generic-components

@eps1lon could we consider dropping these from the return type of forwardRef? It would be a breaking change.

**Reply by johnnyreilly** [reply]
I'd be in favour - the gains seem very much worth it.

**Reply by eps1lon** [reply]
defaultProps, propTypes and displayName are valid properties. We can't disallow them.

**Reply by eps1lon** [reply]
So you need to check if we can drop those properties from the return type and still be able to define them later. 


The priority will always be typing React as it can be used. Not how some people find it convenient. That's how it was done in the past and it led to the types being in such a poor state that takes a long time to fix.
